scipy/spatial/_lloyd_algorithm.py

```diff
@@ -0,0 +1,234 @@
+    hand, L1 and L2 distances should improve. This is especially true with
+    QMC methods which tend to favor the discrepancy over other criteria.
+
+    .. note::
+
+        The current implementation does not intersect the Voronoi Tessellation
+        with the boundaries. This implies that for a low number of points,
+        empirically below 20, no Voronoi cell is touching the boundaries.
+        Hence points cannot be moved close to the boundaries.
+
+        Further improvements could consider the points at infinity so that
+        all boundaries are segments of some Voronoi cells. This would fix
+        the computation of the centroid position.
+
+    .. warning::
+
+       The Voronoi Tessellation step is expensive and quickly becomes
+       intractable with dimensions as low as 10 even for a sample of points
+       of size as low as 1000.
+
+    .. versionadded:: 1.8.0
+
+    References
+    ----------
+    .. [1] Lloyd. "Least Squares Quantization in PCM".
+       IEEE Transactions on Information Theory, 1982.
+    .. [2] Max J. "Quantizing for minimum distortion".
+       IEEE Transactions on Information Theory, 1960.
+
+    Examples
+    --------
+    >>> from scipy.spatial.distance import cdist
+    >>> from scipy.spatial import lloyd_centroidal_voronoi_tessellation
+    >>> rng = np.random.default_rng()
+    >>> points = rng.random((128, 2))
+
+    Compute the quality of the points using the L1 criterion.
+
+    >>> def l1_norm(points):
+    ...    l1 = cdist(points, points, 'cityblock')
+    ...    return np.min(l1[l1.nonzero()])
+
+    >>> l1_norm(points)
+    0.00161...  # random
+
+    Now process the points using Lloyd's algorithm and check the improvement
+    on the L1. The value should increase.
+
+    >>> points = lloyd_centroidal_voronoi_tessellation(points)
+    >>> l1_norm(points)
+    0.0278...  # random
+
+    """
+    points = np.asarray(points).copy()
+
+    if not points.ndim == 2:
+        raise ValueError('Sample is not a 2D array')
+
+    # Checking that points are within the bounds
+    if not ((points.max() <= 1.) and (points.min() >= 0.)):
+        raise ValueError('Sample is out of bounds')
+
+    if qhull_options is None:
+        qhull_options = 'Qbb Qc Qz QJ'
+
+        if points.shape[1] >= 5:
+            qhull_options += ' Qx'
+
+    if decay is None:
+        # Fit an exponential to be 2 at 0 and 1 at `maxiter`.
+        # The decay is used for relaxation.
+        # analytical solution for y=exp(-maxiter/x) - 0.1
+        root = -maxiter / np.log(0.1)
+        decay = [np.exp(-x / root)+0.9 for x in range(maxiter)]
+    else:
+        try:
+            decay = np.broadcast_to(decay, maxiter)  # type: ignore[arg-type]
+        except ValueError as exc:
+            msg = ('decay is not a list of float'
+                   ' of length maxiter')
+            raise ValueError(msg) from exc
+
+    for i in range(maxiter):
+        l1_old = _l1_norm(points=points)
```

Comment:
It looks like if you compute this once _before_ beginning the loop, you can compute it only once per loop iteration instead of twice.

Reply:
Good catch, fixing this.